betty/job/registry.py

```diff
@@ -0,0 +1,59 @@
+"""
+Register jobs for specific context types in advance.
+"""
+
+from __future__ import annotations
+
+from collections import defaultdict
+from typing import final, MutableMapping, MutableSequence, Iterable, Sequence
+
+from typing_extensions import override
+
+from betty.job import Context, JobEntry, _ContextT, Job, _JobP, Runner, Scheduler
+
+
+@final
+class Registry:
+    """
+    Register jobs for context types.
+    """
+
+    def __init__(self):
+        self._jobs: MutableMapping[type[Context], MutableSequence[JobEntry]] = (
+            defaultdict(list)
+        )
+
+    def add(
+        self,
+        context_type: type[_ContextT],
+        job: Job[_ContextT, _JobP],
+        *job_args: _JobP.args,
+        **job_kwargs: _JobP.kwargs,
+    ):
+        """
+        Add a job for the given context type.
+        """
+        self._jobs[context_type].append((job, job_args, job_kwargs))
+
+    def __getitem__(self, context_type: type[_ContextT]) -> Iterable[JobEntry]:
+        yield from self._jobs[context_type]
+
+
+class RegistryRunner(Runner):
+    """
+    Run jobs from registries.
+    """
+
+    def __init__(self, registries: Sequence[Registry]):
+        self._registries = registries
+
+    @override
+    async def run(self, scheduler: Scheduler[_ContextT]) -> None:
+        if len(self._registries) > 1:
```

Comment:
If there is any length at all! Any first entry must be run within the current batch